### PR TITLE
[SPARK-23196] Unify continuous and microbatch V2 sinks

### PR DESCRIPTION
**Pull request by jose-torres**
## What changes were proposed in this pull request?

Replace streaming V2 sinks with a unified StreamWriteSupport interface, with a shim to use it with microbatch execution.

Add a new SQL config to use for disabling V2 sinks, falling back to the V1 sink implementation.

## How was this patch tested?

Existing tests, which in the case of Kafka (the only existing continuous V2 sink) now use V2 for microbatch.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2.scala [hunk 1]
```diff
@@ -62,7 +62,7 @@ case class WriteToDataSourceV2Exec(writer: DataSourceV2Writer, query: SparkPlan)
 
     try {
       val runTask = writer match {
-        case w: ContinuousWriter =>
+        case w: StreamWriter =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2.scala]: Will we hit this branch under micro batch mode?

File: sql/core/src/main/java/org/apache/spark/sql/sources/v2/streaming/MicroBatchWriteSupport.java [hunk 2]
```diff
@@ -1,60 +0,0 @@
...
-   *                incrementing counter representing a consistent set of data; the same batch may
-   *                be started multiple times in failure recovery scenarios, but it will always
-   *                contain the same records.
-   * @param schema the schema of the data to be written.
-   * @param mode the output mode which determines what successive batch output means to this
-   *             sink, please refer to {@link OutputMode} for more details.
-   * @param options the options for the returned data source writer, which is an immutable
-   *                case-insensitive string-to-string map.
-   */
-  Optional<DataSourceV2Writer> createMicroBatchWriter(
```
Comment (cloud-fan) [inline sql/core/src/main/java/org/apache/spark/sql/sources/v2/streaming/MicroBatchWriteSupport.java]: With this API, the implementation can return None if `epochId` is duplicated. How can we achieve this with the new `StreamWriteSupport`?

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala [hunk 3]
```diff
@@ -281,11 +281,9 @@ final class DataStreamWriter[T] private[sql](ds: Dataset[T]) {
         trigger = trigger)
     } else {
       val ds = DataSource.lookupDataSource(source, df.sparkSession.sessionState.conf)
-      val sink = (ds.newInstance(), trigger) match {
-        case (w: ContinuousWriteSupport, _: ContinuousTrigger) => w
-        case (_, _: ContinuousTrigger) => throw new UnsupportedOperationException(
-            s"Data source $source does not support continuous writing")
-        case (w: MicroBatchWriteSupport, _) => w
+      val disabledSources = df.sparkSession.sqlContext.conf.disabledV2StreamingWriters.split(",")
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala]: is this option created for data sources that implement both v1 and v2 APIs?

File: sql/core/src/main/java/org/apache/spark/sql/sources/v2/streaming/MicroBatchWriteSupport.java (hunk 2 above)
Comment (jose-torres) [reply sql/core/src/main/java/org/apache/spark/sql/sources/v2/streaming/MicroBatchWriteSupport.java]: I think the implementation shouldn't do this. It didn't have the option to do so in the V1 API, and it seems hard to reason about data sources second-guessing the engine's decisions about what batches must be run.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2.scala (hunk 1 above)
Comment (jose-torres) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2.scala]: No - the writer will be wrapped in a MicroBatchWriter which doesn't match StreamWriter. I agree this is a bit unclear and would welcome suggestions on how to refactor.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala (hunk 3 above)
Comment (jose-torres) [reply sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala]: Yes. The intent is to support two things:

* Turning off the V2 implementation for a source if someone has problems with it.
* Running some subset of Spark unit tests with V2 disabled, ensuring that the V1 execution pipeline still works for as long as we support it.

File: sql/core/src/main/java/org/apache/spark/sql/sources/v2/streaming/MicroBatchWriteSupport.java (hunk 2 above)
Comment (jose-torres) [reply sql/core/src/main/java/org/apache/spark/sql/sources/v2/streaming/MicroBatchWriteSupport.java]: The writer can drop the duplicate epoch on commit. This is of course less efficient, but the scenario only happens in the rare case where the driver fails between running the writer and dumping to the commit log, so I think that's fine. (The writer will have to implement the drop-on-commit path either way.)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2.scala]: I don't have a better idea, but at least we should add some comments here to explain this.

File: sql/core/src/main/java/org/apache/spark/sql/sources/v2/streaming/MicroBatchWriteSupport.java (hunk 2 above)
Comment (cloud-fan) [reply sql/core/src/main/java/org/apache/spark/sql/sources/v2/streaming/MicroBatchWriteSupport.java]: agreed.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala]: ok so this is only useful for built-in stream sources, as the v1 source API is not public,

File: sql/core/src/main/java/org/apache/spark/sql/sources/v2/streaming/StreamWriteSupport.java [hunk 4]
```diff
@@ -48,7 +48,7 @@
      * @param options the options for the returned data source writer, which is an immutable
      *                case-insensitive string-to-string map.
      */
-    Optional<ContinuousWriter> createContinuousWriter(
+    Optional<StreamWriter> createStreamWriter(
```
Comment (cloud-fan) [inline sql/core/src/main/java/org/apache/spark/sql/sources/v2/streaming/StreamWriteSupport.java]: do we still need to return `Optional`? In which case an implementation should return `None`?

Comment (jose-torres) [reply sql/core/src/main/java/org/apache/spark/sql/sources/v2/streaming/StreamWriteSupport.java]: You're right, we don't need to. Changed to just StreamWriter.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 5]
```diff
@@ -1110,6 +1110,13 @@ object SQLConf {
       .timeConf(TimeUnit.MILLISECONDS)
       .createWithDefault(100)
 
+  val DISABLED_V2_STREAMING_WRITERS = buildConf("spark.sql.streaming.disabledV2Writers")
+    .internal()
+    .doc("A comma-separated list of fully qualified data source register class names for which" +
+      " StreamWriteSupport is disabled. Writes to these sources will faill back to the V1 Sink.")
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: nit. `faill` -> `fail`

File: sql/core/src/main/java/org/apache/spark/sql/sources/v2/streaming/writer/StreamWriter.java [hunk 6]
```diff
@@ -37,8 +41,28 @@
...
+   * happens, or some data writers were committed but their commit messages haven't reached the
+   * driver when the abort is triggered. So this is just a "best effort" for data sources to
+   * clean up the data left by data writers.
+   */
+  void abort(long epochId, WriterCommitMessage[] messages);
+
   default void commit(WriterCommitMessage[] messages) {
     throw new UnsupportedOperationException(
-       "Commit without epoch should not be called with ContinuousWriter");
+       "Commit without epoch should not be called with StreamWriter");
```
Comment (dongjoon-hyun) [inline sql/core/src/main/java/org/apache/spark/sql/sources/v2/streaming/writer/StreamWriter.java]: nit. One more space at the start?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/sources/StreamingDataSourceV2Suite.scala [hunk 7]
```diff
@@ -184,6 +184,24 @@ class StreamingDataSourceV2Suite extends StreamTest {
...
+    // Ensure the V2 path works normally and generates a V2 sink..
+    val v2Query = testPositiveCase(
+      "fake-read-microbatch-continuous", "fake-write-v1-fallback", Trigger.Once())
+    assert(v2Query.asInstanceOf[StreamingQueryWrapper].streamingQuery.sink
+      .isInstanceOf[FakeWriteV1Fallback])
+
+    // Ensure we create a V1 sink with the config. Note the config is a comma separated
+    // list, including other fake entries.
+    val fullSinkName = "org.apache.spark.sql.streaming.sources.FakeWriteV1Fallback"
+    withSQLConf("spark.sql.streaming.disabledV2Writers" -> s"a,b,c,test,$fullSinkName,d,e") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/sources/StreamingDataSourceV2Suite.scala]: "spark.sql.streaming.disabledV2Writers" -> SQLConf.DISABLED_V2_STREAMING_WRITERS.key

**Comment by dongjoon-hyun** [comment]
Could you remove the unused `import` inside the modified files in this PR, too?

**Comment by jose-torres** [comment]
Removed all the ones IntelliJ detected.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 8]
```diff
@@ -1110,6 +1110,13 @@ object SQLConf {
       .timeConf(TimeUnit.MILLISECONDS)
       .createWithDefault(100)
 
+  val DISABLED_V2_STREAMING_WRITERS = buildConf("spark.sql.streaming.disabledV2Writers")
+    .internal()
+    .doc("A comma-separated list of fully qualified data source register class names for which" +
+      " StreamWriteSupport is disabled. Writes to these sources will fail back to the V1 Sink.")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: what will happen if the specified class only implements v2 sink?

File: sql/core/src/main/java/org/apache/spark/sql/sources/v2/streaming/writer/StreamWriter.java [hunk 9]
```diff
@@ -37,8 +41,28 @@
    */
   void commit(long epochId, WriterCommitMessage[] messages);
```
Comment (cloud-fan) [inline sql/core/src/main/java/org/apache/spark/sql/sources/v2/streaming/writer/StreamWriter.java]: shall we also document the requirement about idempotent？

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 8 above)
Comment (jose-torres) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: DataStreamWriter will call DataSource.createSink(), which will notice the providing class doesn't have a (V1) sink implementation and throw "Data source $className does not support streamed writing".

File: sql/core/src/main/java/org/apache/spark/sql/sources/v2/streaming/writer/StreamWriter.java (hunk 9 above)
Comment (jose-torres) [reply sql/core/src/main/java/org/apache/spark/sql/sources/v2/streaming/writer/StreamWriter.java]: Good idea. PTAL at the wording

**Comment by cloud-fan** [comment]
retest this please

**Comment by jose-torres** [comment]
The above failure is an unrelated issue which https://github.com/apache/spark/pull/20398 is out to fix.

**Comment by jose-torres** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
LGTM, pending jenkins

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by cloud-fan** [comment]
thanks, merging to master/2.3!

**Pull request closed without merging**